Moves CollectWgsMetricsFromSampledSites functionality and tests to Co…

meganshand opened the pull request:
…llectWgsMetrics

meganshand commented:
This consolidates CollectWgsMetrics and CollectWgsMetricsFromSampledSites. @yfarjoun Could you please take a look?

nh13 commented:
@yfarjoun this could be a painful rebase for me at #460.  How would you like them ordered.  My PR will be ready shortly.

eitanbanks commented on src/java/picard/analysis/CollectWgsMetrics.java:
```diff
@@ -110,6 +110,9 @@
     @Option(doc="Sample Size used for Theoretical Het Sensitivity sampling. Default is 10000.", optional = true)
     public int SAMPLE_SIZE=10000;
 
+    @Option(shortName = "INTERVALS", doc = "An interval list file that contains the locations of the positions to assess. It is important that the sampled positions be chosen so that they are spread out at least further than a read's length apart; otherwise, you run the risk of double-counting reads in the metrics.", optional = true)
```
Instead of "sampled positions" say "intervals"

eitanbanks commented on src/java/picard/analysis/CollectWgsMetrics.java:
```diff
@@ -350,10 +353,17 @@ protected WgsMetrics generateWgsMetrics() {
+        if(INTERVALS != null){
+            return filter.getFilteredRecords();
+        }
         return filter.getFilteredBases();
```
It's very important to copy over the comment from CollectWgsMetricsFromSampledSites because the behavior is different.  You may even want to mention this in the docs for the INTERVALS argument too.

eitanbanks replied:
Also, this whole method could be shortened to a single line:
return (INTERVALS != null) ? filter.getFilteredRecords() : filter.getFilteredBases();

Oh, and another thing: protocol in this codebase is to put spaces after the 'if' and before the braces.  Otherwise, you risk incurring the wrath of @nh13.

nh13 replied:
exactly @eitanbanks

eitanbanks commented on src/java/picard/analysis/CollectWgsMetrics.java:
```diff
@@ -350,10 +353,17 @@ protected WgsMetrics generateWgsMetrics() {
 
     protected SamLocusIterator getLocusIterator(final SamReader in) {
+        if(INTERVALS != null){
+            IOUtil.assertFileIsReadable(INTERVALS);
```
Same note here about the spacing

nh13 replied:
Could you add this check to the start of the `doWork` method? I think the pattern is to call the `IOUtil` asserts as soon as possible directly in the `doWork` the method.

eitanbanks commented:
Review complete on my end.  @yfarjoun may still want to take a look.

meganshand commented:
I made the changes from @eitanbanks and @nh13. Could someone please take another look?

eitanbanks commented:
This looks really good to me.  @yfarjoun do you want to take a quick look too given that it's a commonly used tool?

nh13 commented:
It would be great if #460 could be merged first, and this rebased on top of that, since this is a much smaller change.

yfarjoun commented:
Yes. let's do that. I'll review #460 later today. thanks for the
understanding @nh13

On Wed, Mar 2, 2016 at 12:33 PM, Nils Homer <notifications@github.com>
wrote:

> It would be great if #460
> <https://github.com/broadinstitute/picard/pull/460> could be merged
> first, and this rebased on top of that, since this is a much smaller change.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/picard/pull/470#issuecomment-191341392>
> .
>

yfarjoun commented:
need to remove line from xml as well.

yfarjoun commented:
since it's deprecated, I'd rather folks not find the tool.

yfarjoun commented on src/java/picard/analysis/CollectWgsMetrics.java:
```diff
@@ -349,12 +358,18 @@ protected WgsMetrics generateWgsMetrics() {
+     */
     protected long getBasesExcludedBy(final CountingFilter filter) {
-        return filter.getFilteredBases();
+        return (INTERVALS != null) ? filter.getFilteredRecords() : filter.getFilteredBases();
```
this only works when "sampled sites" are a single base here and there. This needs to work when INTERVALS is complete contigs...and I think it doesn't. @eitanbanks could you help @meganshand figure out how to make it work...this might be a little more effort than we first thought.

eitanbanks replied:
Yeah, that's hard.  That's why it was called "FromSampledSites".
@yfarjoun what if we check up front and fail if any of the intervals are greater than 1 base?

yfarjoun replied:
but they will be! we want to use this to filter on 1-22,x,y,mt...right?

On Wed, Mar 2, 2016 at 4:58 PM, Eric Banks <notifications@github.com> wrote:

> In src/java/picard/analysis/CollectWgsMetrics.java
> <https://github.com/broadinstitute/picard/pull/470#discussion_r54799033>:
>
> >      protected long getBasesExcludedBy(final CountingFilter filter) {
> > -        return filter.getFilteredBases();
> > +        return (INTERVALS != null) ? filter.getFilteredRecords() : filter.getFilteredBases();
>
> Yeah, that's hard. That's why it was called "FromSampledSites".
> @yfarjoun <https://github.com/yfarjoun> what if we check up front and
> fail if any of the intervals are greater than 1 base?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/picard/pull/470/files#r54799033>.
>

eitanbanks replied:
Oh, duh.  Great point.  You are totally right.
Can we whiteboard this later today?

nh13 replied:
Sorry if I am missing something, but aren't you using a `SamLocusIterator` that returns only records overlapping the intervals, and so you could just return `getFilteredBases` in either case?  What base would be counted that shouldn't?

yfarjoun replied:
The difficulty is that we want this to work when the INTERVALS are isolated loci or large intervals...

yfarjoun commented on src/tests/java/picard/analysis/CollectWgsMetricsTest.java:
```diff
@@ -101,6 +101,66 @@ public void testMetricsFromWGS(final File input, final File outfile, final Strin
+    }
+
+    @Test
+    public void testContiguousIntervals() throws IOException {
```
could you add a test that has two close loci and one far away in the INTERVAL? I think the code will currently fail on that test.

nh13 commented:
@yfarjoun my suggestion was to merge this PR _after_ #460.  My apologies if that was not clear.  Does that still work for you?

yfarjoun commented:
@nh13 my position is unchanged. We will merge #460  first and then #470. Please do not fret. :ok: ?

meganshand commented:
After talking to @eitanbanks, we decided to keep CollectWGSMetricsFromSampledSites and also add an argument to CollectWGSMetrics that will take interval lists with larger intervals such as a whole contig. As long as the LARGE_INTERVALS don't have reads that can overhang an edge it should work. For smaller intervals with overhanging reads CollectWGSMetricsFromSampledSites should still be used. Could you please take another look @yfarjoun?

eitanbanks commented on src/java/picard/analysis/CollectWgsMetrics.java:
```diff
@@ -110,6 +110,12 @@
+    @Option(shortName = "LARGE_INTERVALS", doc = "An interval list file that contains contig-sized locations of the positions to assess. It is important that " +
+            "the intervals be chosen so that they won't have overlapping reads that extend beyond the boundaries of the interval. For smaller sites use " +
+            "CollectWgsMetricsFromSampledSites.",
+            optional = true)
```
I think I would prefer to call this just INTERVALS and change the doc text to something like this:

An interval list file that contains the positions to restrict the assessment.  Please note that all bases of reads that overlap these intervals will be considered, even if some of those bases extend beyond the boundaries of the interval.   The ideal use case for this argument is to use it to restrict the calculation to a subset of (whole) contigs.  To restrict the calculation just to individual positions without overlap, please see CollectWgsMetricsFromSampledSites.

What do you guys think?

meganshand replied:
I think if we want this argument to be INTERVALS, then the argument in CollectWgsMetricsFromSampledSites will need to be something other than INTERVALS. I thought it was better to not change the FromSampledSites version in case there were scripts out there already using it.

eitanbanks replied:
I defer to @yfarjoun

yfarjoun replied:
I think it's fine to use the same argument as long as the doc is clear.

eitanbanks replied:
Note that one is optional and the other is required.

meganshand commented:
I changed the doc to the one suggested by @eitanbanks because I thought it was clearer and I changed the argument name for both tools to INTERVALS. Should be ready for another look, @yfarjoun?

yfarjoun commented:
@h13's PR has been merged.

yfarjoun commented on src/java/picard/analysis/CollectWgsMetrics.java:
```diff
@@ -110,6 +110,13 @@
+            "all bases of reads that overlap these intervals will be considered, even if some of those bases extend beyond the boundaries of " +
+            "the interval. The ideal use case for this argument is to use it to restrict the calculation to a subset of (whole) contigs. To " +
+            "restrict the calculation just to individual positions without overlap, please see CollectWgsMetricsFromSampledSites.",
+            optional = true, overridable = true)
```
why overridable = true?

meganshand replied:
So that it can be overridden by the INTERVALS argument to CollectWgsMetricsFromSampledSites. Without it if both tools have an INTERVALS argument and you get a command line parser error when you run CollectWgsMetricsFromSampledSites.

yfarjoun replied:
got it. thanks!

On Mon, Mar 7, 2016 at 12:52 PM, meganshand <notifications@github.com>
wrote:

> In src/java/picard/analysis/CollectWgsMetrics.java
> <https://github.com/broadinstitute/picard/pull/470#discussion_r55242198>:
>
> > @@ -110,6 +110,13 @@
> >      @Option(doc="Sample Size used for Theoretical Het Sensitivity sampling. Default is 10000.", optional = true)
> >      public int SAMPLE_SIZE=10000;
> >
> > +    @Option(shortName = "INTERVALS", doc = "An interval list file that contains the positions to restrict the assessment. Please note that " +
> > +            "all bases of reads that overlap these intervals will be considered, even if some of those bases extend beyond the boundaries of " +
> > +            "the interval. The ideal use case for this argument is to use it to restrict the calculation to a subset of (whole) contigs. To " +
> > +            "restrict the calculation just to individual positions without overlap, please see CollectWgsMetricsFromSampledSites.",
> > +            optional = true, overridable = true)
>
> So that it can be overridden by the INTERVALS argument to
> CollectWgsMetricsFromSampledSites. Without it if both tools have an
> INTERVALS argument and you get a command line parser error when you run
> CollectWgsMetricsFromSampledSites.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/picard/pull/470/files#r55242198>.
>

yfarjoun commented on src/java/picard/analysis/CollectWgsMetrics.java:
```diff
@@ -349,12 +359,16 @@ protected WgsMetrics generateWgsMetrics() {
 
+    /**
+     * If INTERVALS is specified, this will count bases beyond the interval list when the read overlaps the intervals and extends beyond the
+     * edge. INTERVALS should only include regions that have hard edges without reads that could extend beyond the boundary (such as a whole contig).
```
I would add the word "Ideally" before "INTERVALS"

yfarjoun commented on src/java/picard/analysis/CollectWgsMetricsFromSampledSites.java:
```diff
@@ -46,10 +46,11 @@
                 "It is important that the sampled positions be chosen so that they are spread out " +
                 "at least further than a read's length apart; otherwise, you run the risk of double-counting " +
-                "reads in the metrics.",
+                "reads in the metrics. If a contig-sized interval is needed, use INTERVALS argument in CollectWgsMetrics.",
```
"**if co**ntig-sized interval**s are** needed"

yfarjoun commented on src/tests/java/picard/analysis/CollectWgsMetricsFromSampledSitesTest.java:
```diff
@@ -101,4 +101,36 @@ public void testContiguousIntervals() throws IOException {
+
+        for (final CollectWgsMetrics.WgsMetrics metrics : output.getMetrics()) {
+            Assert.assertEquals(metrics.GENOME_TERRITORY, 404);
+            Assert.assertNotEquals(metrics.PCT_EXC_MAPQ, 0.271403);
```
This is too easy a test to pass. could you subset the "forMetrics.sam" file (temporarily) to only contain contig chr7, run (manually) the normal WGSMetrics and record the results from there? then do an assert*Equals* rather than a *not*equals...

yfarjoun replied:
oh, nm. this is testing CollectWgsMetricsFromSampledSites...

yfarjoun replied:
I'm concerned that it isn't clear where these numbers are coming from. An alternative is to run CollectWGSMetrics here as well and compare the metrics directly, rather than "remembering" the results from the other test.

yfarjoun commented:
- minor changes to text.
- a request for improving the test. 
- you need to rebase, you have conflicts.

meganshand commented:
@yfarjoun I changed the comments and fixed the test as you requested. It's only testing if CollectWgsMetrics returns a different result than CollectWgsMetricsFromSampledSites if you use the exact same inputs (same INTERVALS list).

meganshand commented:
@yfarjoun I added the comment and rebased. Could you take one more quick look to make sure the rebase went smoothly?

yfarjoun commented on src/tests/java/picard/analysis/CollectWgsMetricsFromSampledSitesTest.java:
```diff
@@ -101,4 +106,53 @@ public void testContiguousIntervals() throws IOException {
+        collectWgsMetricsOutput.read(new FileReader(collectWgsOutfile));
+
+        for (final CollectWgsMetrics.WgsMetrics metrics : output.getMetrics()) {
+            Assert.assertEquals(metrics.GENOME_TERRITORY, 404);
```
extract this "404" to a variable up near the definition of the file. (same in the test below)

yfarjoun replied:
ah, nevermind...too petty :-)

yfarjoun commented:
:+1:
